File: app/src/main/java/net/xzos/UpgradeAll/data/RepoDatabase.java
package net.xzos.UpgradeAll.data;

import org.json.JSONException;
import org.json.JSONObject;
import org.litepal.crud.LitePalSupport;

public class RepoDatabase extends LitePalSupport {
    private int id;
    private String name;
    private String api;
    private String url; //  方便用户获得源网址

    private String versionChecker;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getApi() {
        return api;
    }

    public void setApi(String api) {
        this.api = api;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        if (url.substring(url.length() - 1).equals("/")) {
            url = url.substring(0, url.length() - 1);
            // 判断url是否多余 /
        }
        this.url = url;
    }

    public JSONObject getVersionChecker() {
        try {
            return new JSONObject(versionChecker);
        } catch (JSONException e) {
            return null;
        }
    }

    public void setVersionChecker(JSONObject versionChecker) {
        this.versionChecker = versionChecker.toString();
    }

}

File: app/src/main/java/net/xzos/upgradeall/data_manager/database/litepal/RepoDatabase.kt
package net.xzos.upgradeall.data_manager.database.litepal

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import net.xzos.upgradeall.core.data.json.gson.AppConfigGson
import net.xzos.upgradeall.core.data.json.gson.AppDatabaseExtraData
import net.xzos.upgradeall.data_manager.UIConfig
import org.json.JSONObject
import org.litepal.crud.LitePalSupport


internal class RepoDatabase(
        var name: String,
        var url: String,
        var api_uuid: String,
        var type: String
) : LitePalSupport() {
    val id: Long = 0

    private var extra_data: String? = null
    private var versionChecker: String? = null
    private val mutex = Mutex()

    var extraData: AppDatabaseExtraData?
        set(value) {
            runBlocking {
                mutex.withLock {
                    if (value != null)
                        extra_data = Gson().toJson(value)
                }
            }
        }
        get() {
            return if (extra_data != null)
                try {
                    Gson().fromJson(extra_data, AppDatabaseExtraData::class.java)
                } catch (e: JsonSyntaxException) {
                    val json = JSONObject(extra_data!!)
                    val cloudAppConfigStr = json.getString("cloud_app_config")
                    json.remove("cloud_app_config")
                    val extra_data = json.toString()
                    val cloudAppConfig = Gson().fromJson(cloudAppConfigStr, AppConfigGson::class.java)
                    Gson().fromJson(extra_data, AppDatabaseExtraData::class.java).apply {
                        this.cloudAppConfig = cloudAppConfig
                    }.also {
                        extraData = it
                        this.save()
                    }
                }
            else null
        }
    var targetChecker: AppConfigGson.AppConfigBean.TargetCheckerBean?
        get() {
            return if (versionChecker != null) Gson().fromJson(
                    versionChecker, AppConfigGson.AppConfigBean.TargetCheckerBean::class.java
            )
            else null
        }
        set(value) {
            if (value != null)
                versionChecker = Gson().toJson(value)
        }

    override fun save(): Boolean =
            if (name.isNotBlank() && url.isNotBlank() && api_uuid.isNotBlank() && type.isNotBlank())
                super.save()
            else false

    companion object {
        @Transient
        internal const val APP_TYPE_TAG = UIConfig.APP_TYPE_TAG

        @Transient
        internal const val APPLICATIONS_TYPE_TAG = UIConfig.APPLICATIONS_TYPE_TAG
    }
}
